### ISSUE TITLE
Somehow discourage (or embrace!!) Preview Debugger Extensions from public usage

### ISSUE THREAD

**Issue by TomFinley** [dotnet/machinelearning#2977]
We have these debugger extensions meant to make "debugger level" inspection of things like `IDataView` and `IDataLoader` and such like more convenient. How nice!

https://github.com/dotnet/machinelearning/blob/87ede9ef39c10b8511fc7f14f97d1250bbd2b951/src/Microsoft.ML.Data/DebuggerExtensions.cs#L15

Now, this is fine, or would be, except that, wouldn't you know it, people are actually using it in their applications, as opposed to more scalable solutions like the as-enumerable helpers or even direct usage of `IDataView` directly. Which, of course, is bad. 😄 So we should discourage its use as bad practice in the API.

Two ways that occur to me:

1. `Obsolete` attribute on the class to "hint" at people that they shouldn't be doing it (hopefully it will not show up as obsolete in immediate window),

2. Write something to the code analyzer forbidding its use in a project, while allowing it in the immediate window.

Maybe there's some other way.

**Comment by jwood803** [comment]
Perhaps the docs and examples can be updated to explicitly state this as well?

**Comment by TomFinley** [comment]
Yes, we really ought to add the docs the criteria @jwood803 ... I'm starting though to think about whether I'm right though.

I talked a bit with @CESARDELATORRE and it seems like this actually serves a very useful need. So we have the `IDataView` interface itself, which is *efficient* but somewhat verbose to use. We have the convenience of things like `AsEnumerable`, but this requires that there be an actual type to reflect into. Something like preview -- little better really than a glorified `Dictionary<string, object>` -- is a *terrible* API from an efficiency perspective, but maybe with appropriate warnings and education, it is "OK" to have as part of the public API? We don't really have anything "like" it, and @CESARDELATORRE says, "look, if you just want to peek at the results in an ad hoc fashion, is it really so awful?" (Paraphrasing, but, that was the gist.)

So maybe we actually just have it. We don't necessarily encourage it, but it does seem to serve a useful purpose.

**Comment by rosebyte** [comment]
The `DebuggerExtensions` documentation now explicitly states that the class and all its methods are for debugging only and should not be used in production code. This guidance was added in #3446
